yktrace/src/lib.rs

```diff
@@ -210,14 +212,19 @@ impl CompiledTrace {
     /// Create a `CompiledTrace` from a pointer to an array containing: the pointer to the compiled
     /// trace, the pointer to the stackmap, and the size of the stackmap.
     pub fn new(code_ptr: *const c_void) -> Self {
-        let slice = unsafe { slice::from_raw_parts(code_ptr as *const usize, 3) };
+        let slice = unsafe { slice::from_raw_parts(code_ptr as *const usize, 4) };
         let funcptr = slice[0] as *const c_void;
         let smptr = slice[1] as *const c_void;
         let smsize = slice[2] as usize;
+        let aotvals = slice[3] as *mut c_void;
+        // We heap allocated this array in ykllvmwrap to pass the data here. Now that we've
+        // extracted it we no longer need to keep the array around.
+        unsafe { libc::free(code_ptr as *mut c_void) };
```

Comment:
I'm not clear on this part.

Is the intention to free the AOTLiveMap array?

I think some of the confusion stems from the fact that `code_ptr` appears to be more than a pointer to code.

Reply:
Yeah, that variable name is awful. I dunno why it's called that tbh. This doesn't free the AOTLiveMap. It just frees a struct we made in ykllvmwrap to reduce the amount of arguments we need to pass to deopt. The AOTLiveMap can only be freed when the trace dies.

Reply:
Does that struct, the one used for argument reduction, really need to be heap allocated?

Shall we rename this variable then?

Reply:
Unfortunately, it does. Though what I said above wasn't fully correct. It's not used to reduce arguments. It's used to return various things from the trace compiler back to Rust (https://github.com/ykjit/yk/blob/master/ykllvmwrap/src/ykllvmwrap.cc#L224), so we can construct a `CompiledTrace` from it. Once we've constructed the `CompiledTrace` the struct is no longer needed. At least it only lives a short time and once we got the redesigned `HotLocation` up, I imagine we would instead pass in pointers which we then write to.

Reply:
Maybe later we should pass the struct by value, but let's leave it for now. Possibly add a `FIXME`?

Reply:
There's already a fixme here: https://github.com/ykjit/yk/blob/master/ykllvmwrap/src/ykllvmwrap.cc#L224